File: benchmarks/NSubstitute.Analyzers.Benchmarks.Shared/Models/Foo.cs
using NSubstitute.Core;

namespace NSubstitute.Analyzers.Benchmarks.Shared.Models
{
    public class Foo : IFoo
    {
        public void VoidReturningMethod()
        {
        }

        public IFoo ObjectReturningMethod() => null;

        public ConfiguredCall CallInfoReturningMethod()
        {
            return null;
        }

        public ConfiguredCall ConfiguredCallReturningProperty { get; }

        public ConfiguredCall this[int a, int b, int c, int d] => null;

        public virtual void VoidReturningVirtualMethod()
        {
        }

        public void VoidReturningMethodWithArguments(int x, int y, decimal z)
        {
        }

        public IFoo ObjectReturningMethodWithArguments(int x, int y, decimal z) => null;

        public IFoo ObjectReturningMethodWithRefArguments(ref int x, ref int y, ref decimal z) => null;

        public int Property { get; }

        public IFoo ObjectReturningProperty { get; }

        public int this[int i] => 1;

        public IFoo this[int i, int y] => null;

        internal virtual IFoo InternalObjectReturningMethod() => null;

        internal virtual IFoo InternalObjectReturningProperty => null;

        internal virtual IFoo this[int x, int y, int z] => null;
    }
}

File: benchmarks/NSubstitute.Analyzers.Benchmarks.Source.VisualBasic/Models/Foo.vb
Imports System
Imports NSubstitute.Core
Imports System.Threading.Tasks

Namespace Models
    Public Class Foo
        Implements IFoo

        Public Sub VoidReturningMethod() Implements IFoo.VoidReturningMethod
        End Sub

        Public Sub VoidReturningMethodWithArguments(a As Integer, b As Integer, c As Decimal) Implements IFoo.VoidReturningMethodWithArguments
        End Sub
        
        Public Function TaskReturningAsyncMethod() As Task
            Return Task.CompletedTask
        End Function

        Public Function GenericTaskReturningAsyncMethod() As Task(Of Object)
            Return Task.FromResult(New Object())
        End Function
        
        Public Function ObjectReturningMethod() As IFoo Implements IFoo.ObjectReturningMethod
            Return Nothing
        End Function
        
        Public Function ObjectReturningMethodWithArguments(a As Integer, b As Integer, c As Decimal) As IFoo Implements IFoo.ObjectReturningMethodWithArguments
            Return Nothing
        End Function

        Public Function ObjectReturningMethodWithArguments(a As Action(Of Integer), b As Integer, c As Decimal) As IFoo
            Return Nothing
        End Function
        
        Public Function ObjectReturningMethodWithRefArguments(ByRef a As Integer, ByRef b As Integer, ByRef c As Decimal) As IFoo Implements IFoo.ObjectReturningMethodWithRefArguments
            Return Nothing
        End Function
        
        Friend Overridable Function InternalObjectReturningMethod() As IFoo
            Return Nothing
        End Function

        Friend Overridable Function InternalObjectReturningMethodWithArguments(ByVal a As Integer) As IFoo
            Return Nothing
        End Function
        
        Friend Overridable ReadOnly Property InternalObjectReturningProperty As IFoo
            Get
                Return Nothing
            End Get
        End Property
        
        Public Function CallInfoReturningMethod() As ConfiguredCall Implements IFoo.CallInfoReturningMethod
            Return Nothing
        End Function

        Public ReadOnly Property ConfiguredCallReturningProperty As ConfiguredCall Implements IFoo.ConfiguredCallReturningProperty

        Default Public ReadOnly Property Item(a As Integer, b As Integer, c As Integer, d As Integer) As ConfiguredCall Implements IFoo.Item
            Get
                Return Nothing
            End Get
        End Property
        
        Public Property IntReturningProperty As Integer Implements IFoo.IntReturningProperty
        Public ReadOnly Property ObjectReturningProperty As IFoo Implements IFoo.ObjectReturningProperty

        Default Public ReadOnly Property Item(a As Integer) As Integer Implements IFoo.Item
            Get
                Return Nothing
            End Get
        End Property

        Default Public ReadOnly Property Item(a As Integer, b As Integer) As IFoo Implements IFoo.Item
            Get
                Return Nothing
            End Get
        End Property
        
        Default Friend Overridable ReadOnly Property Item(ByVal a As Integer, ByVal b As Integer, ByVal c As Integer) As IFoo
            Get
                Return Nothing
            End Get
        End Property
        
    End Class
End Namespace
